### ISSUE TITLE
Super-gaussian performance

### ISSUE THREAD

**Issue by Joseph459459** [aryafallahi/mithra#66]
I noticed that going from a super-gaussian beam of order 1 to order 10 there is an estimated 10x increase in simulation run-time. 

I am wondering if the bottlenecks are in line 836 of classes.cpp and line 243 of beam.cc, where the orders of the super-gaussian are looped over.

Can this be made for efficient by some kind of pre-computation, so less looping is needed?

**Comment by aryafallahi** [comment]
Hello Joseph,

Sorry for my late reply. I saw your message just yesterday.

Yes, the bottleneck is in the 100 Gaussian functions that need to be calculated for a super-Gaussian beam of order 10. Since this needs to be done for all particles at each time step, it easily scales up to a large computation cost.

I have checked several tricks including using a lookup table but it seemed at the end the direct calculation was the fastest. The solution that I have not checked still is templating the order. This may transfer a part of calculation to the build process and relaxes the computation cost. I need to first check it before committing it and of course if it works.

But, for the time this is the status and yes super-gaussian beams are hard to calculate.

Regards,

Arya
